### ISSUE TITLE
Better feed priorities

### ISSUE THREAD

**Issue by lapineige** [FreshRSS/FreshRSS#1323]
Hello FreshRSS team :)

I have a few feed categories where there are a lot of articles published each day.
This means the other feeds are easily "hidden"/downed between these articles.
However I don't want to exclude them from the main page (and having to exclude all of them, and reading all these categories from time to time).

Would it be possible to have a kind of filter, to exclude a category (or several) from the main page, at least as a global parameters (I could turn it on and off when needed, even if it's not super fast) or at best as a kind of "hide icon" that I could click on to turn off the category ? (I mean, in the category list, you could click on the category to show only this one, or on this icon to hide this one)

Are you interested in implementing this feature ?
I guess it could be made as an extension, but I don't see any way to do it without touching to the PHP code, and I am not knowledgeable enough in PHP to do it.
However I could help on the client side, and to test it :)

**Comment by Alkarex** [comment]
Hello @lapineige 
Each feed has a `priority` attribute, in database. Currently, we only use "10" for showing the feed in the main view, and "0" for not showing it. On the back-end side, implementing what you want would not be very difficult at feed level (not at category level). However, the client side (Web UI) is more challenging: We need to find a good approach for adding one more option to the UI without making it too cluttered, and take care of the translations in the different languages as well as the different CSS themes and also in mobile view.

If you are willing to do the client side, I am happy to implement the changes in the back-end.

The first step is to find a UI approach, which we can agree upon.

I see a few options: adding a new radio button like the four we already have, or add a more interesting/powerful drop down menu allowing to select more precisely a priority (e.g. "important feeds (10)", "normal priority feeds (5)", "all feeds (0)")

![priority](https://cloud.githubusercontent.com/assets/1008324/19429529/762378d0-9450-11e6-9331-b4546ef80def.png)

But both those options would add some clutter for some niche use-cases.

What I think would be more valuable is a two-step approach:
1. Implement a small UI for the search field https://github.com/FreshRSS/FreshRSS/issues/1189 including a "priority" parameter;
2. Use the generated search to make a _user filter_ (in the existing drop-down menu at the right-hand side) https://github.com/FreshRSS/FreshRSS/pull/498

**Comment by lapineige** [comment]
> Each feed has a priority attribute, in database. Currently, we only use "10" for showing the feed in the main view, and "0" for not showing it. On the back-end side, implementing what you want would not be very difficult at feed level (not at category level).

Well it's even better for feeds, I can tweak it with even more precision. (even if it's not super handy in my case, I have ~ 80 feeds to change, but it's ok)

> We need to find a good approach for adding one more option to the UI without making it too cluttered, and take care of the translations in the different languages as well as the different CSS themes and also in mobile view.

I do agree: I might do some mock-ups, because it's not easy to find a way to integrate that without cluttering the UI.

I see 3 options:
- as you said, add an icon next to the others: I think it's too much for such "small" (niche) feature, and it makes a lot of clutter especially on phones.
- create a more powerful drop-down menu: well, if we have other feature to add here, why not, as it doesn't add more mess, stay easy to use, not hard to discover
- just create a filter for priority, that the user can call using the filter menu (and add as a predefined filter (well, I didn't figure out how to do it :/)): it's hidden, only used by advanced users who know it's doable (well, we could add it to the docs, but it's still hidden) but it avoid any clutter

And maybe adding a new preference, to activate this capability ? So if the "normal" user don't need it, there's no clutter, no extra button - but we can add it for the others.

> (e.g. "important feeds (10)", "normal priority feeds (5)", "all feeds (0)")

That would be perfect :-)

**Comment by lapineige** [comment]
Which option seems to be the most interesting in your opinion ?

In my point-of-view it depends of the "need" of this feature: if it's really a niche case, I would go for option 3, but as we still have to implement the check-box for marking the feed as "high priority" in the UI (which result in a new check-box, used for "nothing", so not understood by the user), I would go for option 2.

Do you want me to make some mock-up of this drop-down menu ?
I'm thinking that it could just be a predefined filter (like a built-in "user" filter), that we could select directly.

**Comment by lapineige** [comment]
I have another idea to use this feature, even without excluding feeds: changing the CSS style for high priority feeds. Like making them bold, red, ...
In my case it would be used to improve readability (by highlighting those feeds, are making the "normal one" visually less important) but I guess it could be used to highlight high priority information.

As soon as the feeds are different (like a different class/style, and I think it could be the case with this feat) it would be just a theme change, easily changeable by the users, and allowing different possibilities without touching to the main code.
(Example: in my case I would reduce normal feed visual impact, but someone searching to highlight high priority info would increase it for high priority feeds).

However, it would break the actual themes if both feeds category have a different style, so I would keep the current name for normal feeds, and add another  style property (used for the CSS) for the high priority feeds.

What do you think about it ? :smiley: 

---

Apart from that, I just want to inform you that I have some free time in the next ~ days to work on this (and do to some testing), so don't hesitate to ask me during this time for mock-ups/code/testing ;)

**Comment by Alkarex** [comment]
Hello @lapineige 
Very good :-)

The CSS styling would be easy to do: we could expose the priority for each entry as a `data-p="10"` parameter.
For almost all of the approaches we have discussed, there is a need of having a new option in the configuration of the feeds, to add this higher granularity.

![afficher](https://cloud.githubusercontent.com/assets/1008324/19628297/e78626f8-995a-11e6-9e1d-66577cbaf211.png)

So the checkbox "display in main view" must either be replaced by a drop-down menu or we need to add another checkbox. I believe the drop-down menu would be better, with something like:

ˇˇˇˇˇˇˇˇˇˇˇˇˇˇˇˇˇˇ
- Hidden in main view (priority 0)
- Normal priority (10)
- High priority (20)

ˇˇˇˇˇˇˇˇˇˇˇˇˇˇˇˇˇˇ

Excellent, if you have a bit of time to do some implementation :-)

I suggest the following plan (some steps can be done in parallel):
1. Make the above feed config UI
2. Best guess for the translations of step 1 in the different languages
3. Add HTML `data-p="10"` attribute to the entries
4. Make some quick mock-ups to help deciding whether we go for additional radio button or take advantage of the drop-down menu
5. Make some small changes to the different CSS themes for high-priority feeds
6. Implement the priority filter in the back-end (I will do that)
7. Implement the UI decided from step 4

**Comment by Alkarex** [comment]
Later, we could consider adding "Hidden in own category" with a lower priority than "Hidden in main view", such as -10.

**Comment by lapineige** [comment]
For information, I did not have time to work on this, but I'm still thinking about it, and I'm going to come back later with some material :)

**Comment by aledeg** [comment]
See #1750 for the lower priority.

**Comment by mdemoss** [comment]
I also have a problem with more frequently updating feeds crowding out less frequently updating ones. I'm currently locally testing a _**shuffle**_ feature. I'll reference this issue when I get around to opening a new issue for shuffle. At the moment it draws the top 3 from each feed before shuffling, but I could imagine implementing something like priority by drawing more or less from each feed.

**Comment by mdemoss** [comment]
I opened this PR for _**shuffle sort**_ https://github.com/FreshRSS/FreshRSS/pull/2463

**Comment by math-GH** [comment]
@lapineige What is the status here? Could you please check the newest version of FreshRSS and please summarize what is still missed?

**Comment by lapineige** [comment]
I don't really know nor remember what was my point at that time… :sweat_smile:
I don't really have time to dig into this, so I will close it and maybe reopen one day if needed :)
